File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/common/ClearableSynchronizedPool.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.common;

import androidx.core.util.Pools;
import com.facebook.infer.annotation.Nullsafe;
import javax.annotation.Nullable;

/**
 * Like {@link androidx.core.util.Pools.SynchronizedPool} with the option to clear the pool (e.g. on
 * memory pressure).
 */
@Nullsafe(Nullsafe.Mode.LOCAL)
public class ClearableSynchronizedPool<T> implements Pools.Pool<T> {

  private final Object[] mPool;
  private int mSize = 0;

  public ClearableSynchronizedPool(int maxSize) {
    mPool = new Object[maxSize];
  }

  @Override
  public synchronized @Nullable T acquire() {
    if (mSize == 0) {
      return null;
    }
    mSize--;
    final int lastIndex = mSize;
    T toReturn = (T) mPool[lastIndex];
    mPool[lastIndex] = null;
    return toReturn;
  }

  @Override
  public synchronized boolean release(T obj) {
    if (mSize == mPool.length) {
      return false;
    }
    mPool[mSize] = obj;
    mSize++;
    return true;
  }

  public synchronized void clear() {
    for (int i = 0; i < mSize; i++) {
      mPool[i] = null;
    }
    mSize = 0;
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/common/ClearableSynchronizedPool.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.common

import androidx.core.util.Pools.Pool

/**
 * Like [androidx.core.util.Pools.SynchronizedPool] with the option to clear the pool (e.g. on
 * memory pressure).
 */
internal class ClearableSynchronizedPool<T : Any>(maxSize: Int) : Pool<T> {

  private val pool: Array<Any?> = arrayOfNulls(maxSize)
  private var size = 0

  @Synchronized
  @Suppress("UNCHECKED_CAST", "KotlinGenericsCast")
  override fun acquire(): T? {
    if (size == 0) {
      return null
    }
    size--
    val lastIndex = size
    val toReturn = pool[lastIndex] as T
    pool[lastIndex] = null
    return toReturn
  }

  @Synchronized
  override fun release(instance: T): Boolean {
    if (size == pool.size) {
      return false
    }
    pool[size] = instance
    size++
    return true
  }

  @Synchronized
  fun clear(): Unit {
    for (i in 0 until size) {
      pool[i] = null
    }
    size = 0
  }
}
